File: hooks/format_code.py
#!/usr/bin/env python3

import os
import sys
import subprocess
import json
from pathlib import Path


def get_file_extension(file_path):
    """Get the file extension from a file path."""
    return Path(file_path).suffix.lower()


def format_python(file_path):
    """Format Python files using ruff."""
    try:
        subprocess.run(["ruff", "format", file_path], check=True, capture_output=True)
        return True
    except (subprocess.CalledProcessError, FileNotFoundError):
        return False


def format_rust(file_path):
    """Format Rust files using rustfmt."""
    try:
        subprocess.run(["rustfmt", file_path], check=True, capture_output=True)
        return True
    except (subprocess.CalledProcessError, FileNotFoundError):
        return False


def format_go(file_path):
    """Format Go files using gofmt."""
    try:
        subprocess.run(["gofmt", "-w", file_path], check=True, capture_output=True)
        return True
    except (subprocess.CalledProcessError, FileNotFoundError):
        return False


def format_javascript_typescript(file_path):
    """Format JS/TS files using prettier if available, fallback to built-in tools."""
    # Try prettier first
    try:
        subprocess.run(
            ["prettier", "--write", file_path], check=True, capture_output=True
        )
        return True
    except (subprocess.CalledProcessError, FileNotFoundError):
        pass

    # Fallback to eslint --fix if available
    try:
        subprocess.run(["eslint", "--fix", file_path], check=True, capture_output=True)
        return True
    except (subprocess.CalledProcessError, FileNotFoundError):
        return False


def should_format_file(file_path, tool_name):
    """Check if file should be formatted based on tool name."""
    if not os.path.exists(file_path):
        return False
    # Only format if the file was actually modified by Write, Edit, or MultiEdit tools
    if tool_name not in ["Write", "Edit", "MultiEdit"]:
        return False
    return True


def main():
    """Main formatting function."""
    # Read tool data from stdin
    hook_input = sys.stdin.read().strip()
    if not hook_input:
        sys.exit(0)

    try:
        hook_data = json.loads(hook_input)
        tool_name = hook_data.get("tool_name", "")
        tool_input = hook_data.get("tool_input", {})
        file_path = tool_input.get("file_path")

        if not file_path or not should_format_file(file_path, tool_name):
            sys.exit(0)

        extension = get_file_extension(file_path)
        formatted = False
        if extension == ".py":
            formatted = format_python(file_path)
        elif extension == ".rs":
            formatted = format_rust(file_path)
        elif extension == ".go":
            formatted = format_go(file_path)
        elif extension in [".js", ".ts", ".jsx", ".tsx"]:
            formatted = format_javascript_typescript(file_path)

        if formatted:
            print(f"Formatted {file_path}")
    except (json.JSONDecodeError, KeyError):
        # Silently ignore if we can't parse tool args
        pass
    sys.exit(0)


if __name__ == "__main__":
    main()

File: hooks/format-code.js
#!/usr/bin/env node
/**
 * Format Code - PostToolUse Hook for Write|Edit
 * Auto-formats files after Claude Code modifies them.
 * Supports Python (ruff) and Markdown/YAML/JSON (prettier).
 * Logs to: ~/.claude/hooks-logs/
 *
 * Benefits:
 *   - No manual formatting needed after Claude edits files
 *   - Consistent code style across AI-generated changes
 *   - Supports Python and common config/documentation formats
 *
 * Setup in .claude/settings.json:
 * {
 *   "hooks": {
 *     "PostToolUse": [{
 *       "matcher": "Write|Edit",
 *       "hooks": [{ "type": "command", "command": "node /path/to/format-code.js" }]
 *     }]
 *   }
 * }
 */

const fs = require('fs');
const path = require('path');
const { spawnSync } = require('child_process');

const LOG_DIR = path.join(process.env.HOME, '.claude', 'hooks-logs');

const FORMATTERS = {
  '.py': {
    name: 'ruff',
    commands: (fp) => [
      ['uv', 'run', 'ruff', 'check', '--fix', '--exit-zero', '--quiet', fp],
      ['uv', 'run', 'ruff', 'format', '--quiet', fp],
    ],
  },
  '.md':    { name: 'prettier', commands: (fp) => [['npx', '--yes', 'prettier', '--write', fp]] },
  '.yaml':  { name: 'prettier', commands: (fp) => [['npx', '--yes', 'prettier', '--write', fp]] },
  '.yml':   { name: 'prettier', commands: (fp) => [['npx', '--yes', 'prettier', '--write', fp]] },
  '.json':  { name: 'prettier', commands: (fp) => [['npx', '--yes', 'prettier', '--write', fp]] },
};

function log(data) {
  try {
    if (!fs.existsSync(LOG_DIR)) fs.mkdirSync(LOG_DIR, { recursive: true });
    const file = path.join(LOG_DIR, `${new Date().toISOString().slice(0, 10)}.jsonl`);
    fs.appendFileSync(file, JSON.stringify({ ts: new Date().toISOString(), hook: 'format-code', ...data }) + '\n');
  } catch {}
}

function getFormatter(filePath) {
  const ext = path.extname(filePath).toLowerCase();
  return FORMATTERS[ext] || null;
}

function formatFile(filePath, cwd, sessionId) {
  const absPath = path.isAbsolute(filePath) ? filePath : path.join(cwd || process.cwd(), filePath);

  if (!fs.existsSync(absPath)) {
    log({ level: 'SKIP', reason: 'file not found', file: absPath, session_id: sessionId });
    return { success: false, error: 'file not found' };
  }

  const formatter = getFormatter(absPath);
  if (!formatter) {
    log({ level: 'SKIP', reason: 'unsupported file type', file: absPath, session_id: sessionId });
    return { success: false, error: 'unsupported file type' };
  }

  const dir = path.dirname(absPath);

  for (const args of formatter.commands(absPath)) {
    try {
      const result = spawnSync(args[0], args.slice(1), { cwd: dir, stdio: 'pipe' });
      if (result.status !== 0) {
        const msg = result.stderr?.toString()?.trim() || `Process exited with code ${result.status}`;
        log({ level: 'ERROR', formatter: formatter.name, file: absPath, error: msg, session_id: sessionId });
        return { success: false, error: msg, formatter: formatter.name };
      }
    } catch (e) {
      const msg = e.message;
      log({ level: 'ERROR', formatter: formatter.name, file: absPath, error: msg, session_id: sessionId });
      return { success: false, error: msg, formatter: formatter.name };
    }
  }

  log({ level: 'FORMATTED', formatter: formatter.name, file: absPath, session_id: sessionId });
  return { success: true, formatter: formatter.name };
}

async function main() {
  let input = '';
  for await (const chunk of process.stdin) input += chunk;

  try {
    const data = JSON.parse(input);
    const { tool_name, tool_input, session_id, cwd } = data;

    if (!['Write', 'Edit'].includes(tool_name)) {
      return console.log('{}');
    }

    const workDir = cwd || process.cwd();

    if (!tool_input?.file_path) {
      log({ level: 'SKIP', reason: 'no file_path', tool: tool_name, session_id });
      return console.log('{}');
    }

    const files = [tool_input.file_path];

    for (const file of files) {
      formatFile(file, workDir, session_id);
    }

    console.log('{}');
  } catch (e) {
    log({ level: 'ERROR', error: e.message });
    console.log('{}');
  }
}

if (require.main === module) {
  main();
} else {
  module.exports = { getFormatter, formatFile, log, FORMATTERS };
}
